In [45]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn

# Carga de datos

In [24]:
datos = pd.read_csv("./california_housing.csv")

In [25]:
datos

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

# Análisis exploratorio

## Datos numéricos

In [26]:
# Resporte estadístico de los datos cargados
datos.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

Vemos que hay un problema con count de total_bedrooms ya que faltan datos en comparación con el resto de los datos del mismo count.

## Datos categóricos

In [27]:
# Obtenemos los valores únicos de la columna ocean_proximity
datos.ocean_proximity.unique()

# Es lo mismo que
# datos['ocean_proximity'].unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [28]:
# Verificar datos sin valor
datos.isna().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

Vemos que en el total_bedroom hay 207 valores que no tienen valor asignado.

# Limpieza de datos

Eliminar items con elementos faltantes

Alternativamente, se puede realizar imputación de valores

In [29]:
# Borramos los valores nulos
datos.dropna(inplace=True)

# Con inplace=True hacemos que los nuevos datos con las filas vacías dropeadas
# se guarden en la misma variable

# El inplace=True es lo mismo que hacer
# datos = datos.dropna()

In [30]:
datos.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
dtype: int64

Vemos que haciendo uso de la función dropna(inplace=True) eliminamos los elementos sin valor de la data.

# One-hot encoding

Cambiamos los datos categóricos (texto, sin un orden natural) a nuevas numéricas
con números (0 y 1) y borramos la columna original

## Usando Pandas

In [31]:
tabla_dummies = pd.get_dummies(datos.ocean_proximity)

In [32]:
tabla_dummies

<1H OCEAN  INLAND  ISLAND  NEAR BAY  NEAR OCEAN
0              0       0       0         1           0
1              0       0       0         1           0
2              0       0       0         1           0
3              0       0       0         1           0
4              0       0       0         1           0
...          ...     ...     ...       ...         ...
20635          0       1       0         0           0
20636          0       1       0         0           0
20637          0       1       0         0           0
20638          0       1       0         0           0
20639          0       1       0         0           0

[20433 rows x 5 columns]

In [33]:
# Concatenamos las tablas de los datos y la tabla_dummies
datos_preproc = pd.concat([datos, tabla_dummies], axis=1)

In [34]:
datos_preproc

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  <1H OCEAN  INLAND  ISLAND  NEAR BAY  NEAR OCEAN  
0            NEAR BAY          0       0       0         1           0  
1            NEAR BAY          0       0       0         1           0  
2            NEAR BAY          0       0       0         1           0  
3            NEAR BAY          0       0       0         1           0  
4            NEAR BAY          0       0       0         1           0  
...               ...        ...     ...     ...       ...         ...  
20635          INLAND          0       1       0         0           0  
20636          INLAND          0       1       0         0           0  
20637          INLAND          0       1       0         0           0  
20638          INLAND          0       1       0         0           0  
20639          INLAND          0       1       0         0           0  

[20433 rows x 15 columns]

Es necesario concatenar con axis=1 ya que pandas espera que los nombres de las columnas de las tablas sean las mismas. Si estas no lo son, colocará muchos NA, cosa que no queremos.

In [35]:
# Borramos la columna de ocean_proximity y guardamos esta nueva tabla en la misma variable
datos_preproc.drop(columns='ocean_proximity', inplace=True)

# El inplace=True es lo mismo que hacer
# datos_preproc = datos_preproc.drop(columns='ocean_proximity')

In [36]:
datos_preproc

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  <1H OCEAN  \
0           322.0       126.0         8.3252            452600.0          0   
1          2401.0      1138.0         8.3014            358500.0          0   
2           496.0       177.0         7.2574            352100.0          0   
3           558.0       219.0         5.6431            341300.0          0   
4           565.0       259.0         3.8462            342200.0          0   
...           ...         ...            ...                 ...        ...   
20635       845.0       330.0         1.5603             78100.0          0   
20636       356.0       114.0         2.5568             77100.0          0   
20637      1007.0       433.0         1.7000             92300.0          0   
20638       741.0       349.0         1.8672             84700.0          0   
20639      1387.0       530.0         2.3886             89400.0          0   

       INLAND  ISLAND  NEAR BAY  NEAR OCEAN  
0           0       0         1           0  
1           0       0         1           0  
2           0       0         1           0  
3           0       0         1           0  
4           0       0         1           0  
...       ...     ...       ...         ...  
20635       1       0         0           0  
20636       1       0         0           0  
20637       1       0         0           0  
20638       1       0         0           0  
20639       1       0         0           0  

[20433 rows x 14 columns]

# Separar datos de entrenamiento y testing

In [37]:
# Sacamos una muestra del 80% de los datos preprocesados, con un random_state para tener siempre la misma muestra y no esté cambiando cada vez que carguemos
train = datos_preproc.sample(frac=0.8, random_state=123)

# Tomamos el 100% de los datos preprocesados que no están en el train con el
# mismo random_state de este último
test = datos_preproc.drop(train.index).sample(frac=1.0, random_state=123)

# Es importante que tengan el mismo random_state

In [38]:
train

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
14354    -117.23     32.75                23.0       2415.0           653.0   
12908    -121.29     38.63                24.0       2868.0           527.0   
19545    -120.97     37.63                39.0       2360.0           607.0   
12188    -117.35     33.69                11.0       1229.0           236.0   
14786    -117.12     32.57                21.0       1738.0           295.0   
...          ...       ...                 ...          ...             ...   
13853    -117.31     34.50                14.0       2443.0           447.0   
8048     -118.17     33.84                45.0       1853.0           328.0   
15731    -122.44     37.78                52.0       3017.0           851.0   
10837    -117.94     33.68                26.0       4183.0           539.0   
14146    -117.07     32.74                38.0       1901.0           392.0   

       population  households  median_income  median_house_value  <1H OCEAN  \
14354      1275.0       596.0         3.1389            101800.0          0   
12908      1284.0       487.0         3.3182            213000.0          0   
19545      2047.0       605.0         1.7054             58800.0          0   
12188       581.0       190.0         3.1020            111300.0          1   
14786       983.0       298.0         4.8274            174100.0          0   
...           ...         ...            ...                 ...        ...   
13853       883.0       465.0         2.1111            116700.0          0   
8048        945.0       320.0         5.0787            219200.0          1   
15731      1588.0       800.0         3.3882            471400.0          0   
10837      1504.0       520.0         7.4056            374200.0          1   
14146      1099.0       406.0         2.7661            113900.0          0   

       INLAND  ISLAND  NEAR BAY  NEAR OCEAN  
14354       0       0         0           1  
12908       1       0         0           0  
19545       1       0         0           0  
12188       0       0         0           0  
14786       0       0         0           1  
...       ...     ...       ...         ...  
13853       1       0         0           0  
8048        0       0         0           0  
15731       0       0         1           0  
10837       0       0         0           0  
14146       0       0         0           1  

[16346 rows x 14 columns]

Vemos que ***train*** contiene únicamente el 80% del total de los datos preprocesados

In [39]:
test

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
7584     -118.22     33.90                22.0        312.0           107.0   
5092     -118.28     33.97                35.0       2305.0           634.0   
17734    -121.80     37.30                16.0        906.0           149.0   
3073     -119.23     35.79                31.0       2862.0           606.0   
15424    -117.27     33.20                34.0       1852.0           322.0   
...          ...       ...                 ...          ...             ...   
20536    -121.73     38.55                34.0       1717.0           393.0   
6987     -118.03     33.97                39.0       2126.0           434.0   
17679    -121.84     37.32                22.0       3015.0           581.0   
17603    -121.88     37.29                44.0       2026.0           396.0   
18232    -122.11     37.41                27.0       5110.0          1599.0   

       population  households  median_income  median_house_value  <1H OCEAN  \
7584        583.0       119.0         1.9423             98400.0          1   
5092       1978.0       568.0         1.3750            100000.0          1   
17734       605.0       148.0         4.8173            235600.0          1   
3073       2467.0       600.0         2.3125             62100.0          0   
15424       978.0       332.0         4.3542            156900.0          1   
...           ...         ...            ...                 ...        ...   
20536      1224.0       387.0         2.7917            130800.0          0   
6987       1103.0       433.0         3.2852            196200.0          1   
17679      2491.0       530.0         4.3419            176300.0          1   
17603       908.0       383.0         4.1406            274100.0          1   
18232      2764.0      1482.0         3.4198            351900.0          0   

       INLAND  ISLAND  NEAR BAY  NEAR OCEAN  
7584        0       0         0           0  
5092        0       0         0           0  
17734       0       0         0           0  
3073        1       0         0           0  
15424       0       0         0           0  
...       ...     ...       ...         ...  
20536       1       0         0           0  
6987        0       0         0           0  
17679       0       0         0           0  
17603       0       0         0           0  
18232       0       0         1           0  

[4087 rows x 14 columns]

Vemos que ***test*** contiene el 20% restante de los datos.

# Regresión lineal

Se nos pide predecir el valor medio de una cada (median_house_value)

In [40]:
target = 'median_house_value'

1. Para que nuestro modelo predizca esto, primero hay que separar la data de
entrada que la de salida (la respuesta, el target)

- x = variable de entrada
- y = variable de salida (respuesta, el target)

In [46]:
# Varaiable de entrada
x_train = train.drop(columns=target)

# Variable de salida
y_train = train[target]

Hay que hacer esto tanto para la de entrenamiento como la de testing

In [43]:
# Varaiable de entrada
x_test = test.drop(columns=target)

In [44]:
# Variable de salida
y_test = test[target]